From 2b724a547f7fa754173d05b198e01c2071236633 Mon Sep 17 00:00:00 2001
From: Devin Pohl <pohl.devin@gmail.com>
Date: Thu, 28 Oct 2021 14:57:46 -0600
Subject: [PATCH] [PATCH] PCI: Quirk Silicon Motion SM2262/SM2263 NVMe
 controller reset

This is a patch upon a patch. The initial patch has the info:
- https://bugzilla.kernel.org/show_bug.cgi?id=202055
- commit b5a553a6ed8385eb6cafc1f18c7f9d559850a2e1
- Author: Alex Williamson <alex.williamson@redhat.com>

Combined, this commit fixes vfio rebinding for NVME controlers that
use SM2262 and SM2263 controlers. This will allow the vfio-pci driver
to properly rebind when requested and allow PCIE passthrough of these
drives.

Updates have been made recently to reflect upstream work.
---
 drivers/pci/pci.c    |  2 +-
 drivers/pci/pci.h    |  1 +
 drivers/pci/quirks.c | 27 +++++++++++++++++++++++++++
 3 files changed, 29 insertions(+), 1 deletion(-)

diff --git a/drivers/pci/pci.c b/drivers/pci/pci.c
index a4eb0c042ca3..8efd2abb9539 100644
--- a/drivers/pci/pci.c
+++ b/drivers/pci/pci.c
@@ -4995,7 +4995,7 @@ int pci_bridge_secondary_bus_reset(struct pci_dev *dev)
 }
 EXPORT_SYMBOL_GPL(pci_bridge_secondary_bus_reset);
 
-static int pci_parent_bus_reset(struct pci_dev *dev, int probe)
+int pci_parent_bus_reset(struct pci_dev *dev, int probe)
 {
 	struct pci_dev *pdev;
 
diff --git a/drivers/pci/pci.h b/drivers/pci/pci.h
index 93dcdd431072..640d9d26837e 100644
--- a/drivers/pci/pci.h
+++ b/drivers/pci/pci.h
@@ -35,6 +35,7 @@ int pci_mmap_fits(struct pci_dev *pdev, int resno, struct vm_area_struct *vmai,
 
 int pci_probe_reset_function(struct pci_dev *dev);
 int pci_bridge_secondary_bus_reset(struct pci_dev *dev);
+int pci_parent_bus_reset(struct pci_dev *dev, int probe);
 int pci_bus_error_reset(struct pci_dev *dev);
 
 #define PCI_PM_D2_DELAY         200	/* usec; see PCIe r4.0, sec 5.9.1 */
diff --git a/drivers/pci/quirks.c b/drivers/pci/quirks.c
index 8c3c1ef92171..82255ebd70ca 100644
--- a/drivers/pci/quirks.c
+++ b/drivers/pci/quirks.c
@@ -3999,6 +3999,30 @@ static int reset_hinic_vf_dev(struct pci_dev *pdev, int probe)
 	return 0;
 }
 
+
+/*
+ * The Silicon Motion SM2262 is an NVMe controller used by several vendors
+ * which is know to corrupt its MSI-X capability after FLR (PBA offset from
+ * the vector table only allows for 16 entries (0x100 offset) but after FLR
+ * the MSI-X capability reports 22 entries).  As this is a single function
+ * device, a bus reset should offer no significant loss of functionality
+ * versus FLR on this device.  Downstream ports blacklisting bus reset via
+ * PCI_DEV_FLAGS_NO_BUS_RESET will fall back to FLR as we have no workaround
+ * for them.
+ *
+ * Link: https://bugzilla.kernel.org/show_bug.cgi?id=202055
+ *
+ * This controller is known to be used in the Intel 760p (8086:f1a6) and
+ * ADATA XPG SX8200 (126f:2262), the latter making use of the native PCI
+ * vendor and device ID from Silicon Motion.  Also believed to be affected
+ * are the Mushkin Pilot, HP EX920, and Western Digital Black, though it's
+ * not known if any of these override the native device ID as Intel does.
+ */
+static int prefer_bus_reset(struct pci_dev *dev, int probe)
+{
+	return pci_parent_bus_reset(dev, probe);
+}
+
 static const struct pci_dev_reset_methods pci_dev_reset_methods[] = {
 	{ PCI_VENDOR_ID_INTEL, PCI_DEVICE_ID_INTEL_82599_SFP_VF,
 		 reset_intel_82599_sfp_virtfn },
@@ -4009,6 +4033,9 @@ static const struct pci_dev_reset_methods pci_dev_reset_methods[] = {
 	{ PCI_VENDOR_ID_SAMSUNG, 0xa804, nvme_disable_and_flr },
 	{ PCI_VENDOR_ID_INTEL, 0x0953, delay_250ms_after_flr },
 	{ PCI_VENDOR_ID_INTEL, 0x0a54, delay_250ms_after_flr },
+	{ 0x126f, 0x2262, prefer_bus_reset },
+	{ 0x126f, 0x2263, prefer_bus_reset },
+	{ PCI_VENDOR_ID_INTEL, 0xf1a6, prefer_bus_reset },
 	{ PCI_VENDOR_ID_CHELSIO, PCI_ANY_ID,
 		reset_chelsio_generic_dev },
 	{ PCI_VENDOR_ID_HUAWEI, PCI_DEVICE_ID_HINIC_VF,
-- 
2.33.1

